ding/rl_utils/td.py

```diff
@@ -1012,6 +1012,146 @@ def iqn_nstep_td_error(
+    Shapes:
+        - data (:obj:`q_nstep_td_data`): the q_nstep_td_data containing\
+        ['q', 'next_n_q', 'action', 'reward', 'done']
+        - q (:obj:`torch.FloatTensor`): :math:`(B, tau, N)` i.e. [batch_size, tau, action_dim]
+        - next_n_q (:obj:`torch.FloatTensor`): :math:`(B, tau', N)`
+        - action (:obj:`torch.LongTensor`): :math:`(B, )`
+        - next_n_action (:obj:`torch.LongTensor`): :math:`(B, )`
+        - reward (:obj:`torch.FloatTensor`): :math:`(T, B)`, where T is timestep(nstep)
+        - done (:obj:`torch.BoolTensor`) :math:`(B, )`, whether done in last timestep
+        - quantiles_hats (:obj:`torch.FloatTensor`): :math:`(B, tau)`
+    """
+    q, next_n_q, action, next_n_action, reward, done, quantiles_hats, weight = data
+
+    assert len(action.shape) == 1, action.shape
+    assert len(next_n_action.shape) == 1, next_n_action.shape
+    assert len(done.shape) == 1, done.shape
+    assert len(q.shape) == 3, q.shape
+    assert len(next_n_q.shape) == 3, next_n_q.shape
+    assert len(reward.shape) == 2, reward.shape
+
+    if weight is None:
+        weight = torch.ones_like(action)
+
+    batch_size = done.shape[0]
+    tau = q.shape[1]
+    tau_prime = next_n_q.shape[1]
+
+    # shape: batch_size x tau x 1
+    q_s_a = evaluate_quantile_at_action(q, action)
+    # shape: batch_size x tau_prim x 1
+    target_q_s_a = evaluate_quantile_at_action(next_n_q, next_n_action)
+
+    assert reward.shape[0] == nstep
+    device = torch.device("cuda" if reward.is_cuda else "cpu")
+    reward_factor = torch.ones(nstep).to(device)
+    for i in range(1, nstep):
+        reward_factor[i] = gamma * reward_factor[i - 1]
+    reward = torch.matmul(reward_factor, reward)  # [batch_size]
+    if value_gamma is None:
+        target_q_s_a = reward.unsqueeze(-1) + (gamma ** nstep) * target_q_s_a.squeeze(-1) * (1 - done).unsqueeze(-1)
+    else:
+        target_q_s_a = reward.unsqueeze(-1) + value_gamma.unsqueeze(-1) * target_q_s_a.squeeze(-1) * (1 - done
+                                                                                                      ).unsqueeze(-1)
+    target_q_s_a = target_q_s_a.unsqueeze(-1)
+
+    # shape: batch_size x tau' x tau x 1.
+    bellman_errors = (target_q_s_a[:, :, None, :] - q_s_a[:, None, :, :])
+
+    # The huber loss (see Section 2.3 of the paper) is defined via two cases:
+    huber_loss = torch.where(
+        bellman_errors.abs() <= kappa, 0.5 * bellman_errors ** 2, kappa * (bellman_errors.abs() - 0.5 * kappa)
+    )
+
+    # shape: batch_size x num_tau_prime_samples x num_tau_samples x 1.
+    quantiles_hats = quantiles_hats[:, None, :, None].repeat([1, tau_prime, 1, 1])
+
+    abcd = torch.abs(quantiles_hats - ((bellman_errors < 0).float()).detach())
+    # shape: batch_size x tau_prime x tau x 1.
+    quantile_huber_loss = (torch.abs(quantiles_hats - ((bellman_errors < 0).float()).detach()) * huber_loss) / kappa
```

Comment:
if the former is just a kind of weight, wrapper it `torch.no_grad` is better than detach